### ISSUE TITLE
Style guidelines

### ISSUE THREAD

**Issue by tcstewar** [nengo/nengo_gui#684]
Once 0.2 is done, that'd be a great time to do a pass through and fix formatting.  This would require:

- [ ] make sure all the python code passes flake8
- [ ] make sure javascript code passes jscs
- [ ] make sure we have ```from __future__ import print_function``` where needed

**Comment by Seanny123** [comment]
I would also recommend embedding these checks into TravisCI, as Nengo currently does, so that styling is checked before pull requests are merged.

**Comment by tcstewar** [comment]
> I would also recommend embedding these checks into TravisCI, as Nengo currently does, so that styling is checked before pull requests are merged.

That will happen after 0.2.  Right now the entire code base just horribly fails the style checks.  But I want to wait until after the current PRs are merged before cleaning up the code, so as to avoid merge conflicts.

**Comment by Seanny123** [comment]
Now that 0.2 has been merged, is this a priority again? Or are we waiting for a co-op to take this off our hands?

**Comment by tcstewar** [comment]
Yup, it should be a priority again, although I'm trying to get #673 and #746 merged first, as they touch a lot of files.

**Comment by tbekolay** [comment]
Not sure if a JS standard has been chosen, but I started using Standard JS for the turntable code and it's... you know, okay: http://standardjs.com/rules.html No semicolons, so that's a thing. It's a private repo, but [here's the turntable JS](https://github.com/abr/turntable/tree/master/server/static/js).

**Comment by Seanny123** [comment]
A standard was decided upon during a meeting a while ago. Basically, it was decided that we would follow the Google standard for JS, with a modified commenting style to match Python and this was input in the `jscs` file. If everyone wants to switch, that's fine. I don't really feel strongly about JavaScript style. I basically think it looks ugly no matter what you do to it.

**Comment by Seanny123** [comment]
I changed my mind. After re-reading the [Google standard](https://google.github.io/styleguide/javascriptguide.xml?showone=Semicolons#Semicolons) and [the article linked from the StandardJS](http://inimino.org/~inimino/blog/javascript_semicolons), I've decided that I like semi-colons. But if two or more people disagree with me than I'm fine with not using them.

**Comment by tbekolay** [comment]
Ah cool, will set up jscs in my editor then. Thanks!

**Comment by tbekolay** [comment]
Would also be useful to add running jscs and flake8 to `travis.yml`.

**Comment by Seanny123** [comment]
> Would also be useful to add running jscs and flake8 to travis.yml

That's mentioned in OP, if you want to take a shot at it, (since neither Terry, nor I, understand `travis.yml` and it doesn't appear to be a case of copying from `nengo/nengo`) it would be greatly appreciated.

**Comment by tbekolay** [comment]
Another useful thing for the style guide that I noticed reviewing #746 (if we don't already have this somewhere)

- [ ] HTML tag IDs should be all lowercase with hyphens between words rather than underscores.

This is the convention used by most web things that I've seen, and I believe the reason is because IDs can appear in URIs (when you link to something within a page e.g. `mysite.com/index.html#my-header`). URIs are (used to be?) case insensitive hence all lowercase, and some SEO blog I read at some point said that hyphens are better than underscores in URIs so that's probably where that comes from.

> That's mentioned in OP, if you want to take a shot at it it would be greatly appreciated

Ah didn't see it sorry! Yes, I can do that no problem.

**Comment by Seanny123** [comment]
On the topic of HTML and CSS style checking, this is apparently [a giant rabbit hole](http://stackoverflow.com/questions/6646204/are-there-any-command-line-validation-tools-for-html-and-css).

For HTML, the one that seems easiest to install [this mozilla script](https://github.com/mozilla/html5-lint).

For CSS, [csslint](https://github.com/CSSLint/csslint/wiki/command-line-interface) seems pretty straight-forward.

Thoughts?

**Comment by tbekolay** [comment]
For CSS I know of http://stylelint.io/ as being relatively good

**Comment by Seanny123** [comment]
@tbekolay Can I let you take of the CSS/HTML, since you seem to be the only who knows how to do this properly?

**Comment by tbekolay** [comment]
:neckbeard: 👍

**Comment by krisfrenette** [comment]
This is such a big rabbit hole to get sucked into. I would suggest using TSLint and Prettier for all of the client side work. My preference would be to accept what prettier does, and we accept it out of the box. IF there is something that we really don't like then we can look at adding that to the config, but I would strongly advise against it. Doing all of this will allow us to be more effective when we are submitting our PR's and we can then focus on what is important. 

Having said that we probably should look at all of this in the TS rewrite and apply it then.
